Move reorder windows optimizer to the new iterative optimizer

maciejgrzybek opened the pull request:
First pass was here: https://github.com/Teradata/presto/pull/493

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsByPartitionsOrder.java:
```diff
@@ -0,0 +1,90 @@
+
+    private static boolean dependsOn(WindowNode parent, WindowNode dependent)
+    {
+        return parent.getPartitionBy().stream().anyMatch(symbol -> dependent.getCreatedSymbols().contains(symbol))
```
Add unit tests for this branch

maciejgrzybek replied:
Added in: ca4e7c77ce1398409f7fe0f5eea41b9568b24203

kokosing replied:
Why not to use `DepenendencyExtractor` somehow here? This code is complicated and seems to be fragile. When `WindowNode` internal structure changes this would need to be updated, in some cases it might be not trivial.

I would write something like this below:
```
Set<Symbol> parentDepenencies = DependencyExtractor.extractUnique(paranent);
return dependant.getOutputSymbols().stream().anyMatch(parentDependencies::contains);
```

maciejgrzybek replied:
@kokosing, please have a look here: https://github.com/prestodb/presto/pull/7298#discussion_r100112154

maciejgrzybek replied:
Similar solution applied here: e585eb85154b3ef50aaa9702e4aa4744958cc97c

kokosing replied:
and similar comments here (https://github.com/prestodb/presto/pull/7298#discussion_r100478109)

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsByPartitionsOrder.java:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+
+    private static boolean dependsOn(WindowNode parent, WindowNode dependent)
```
s/dependent/child

as it it is not know yet if they depended on each other

maciejgrzybek replied:
Fixed in e585eb85154b3ef50aaa9702e4aa4744958cc97c.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestSwapAdjacentWindowsByPartitionsOrder.java:
```diff
@@ -0,0 +1,107 @@
+import static com.facebook.presto.sql.tree.FrameBound.Type.CURRENT_ROW;
+import static com.facebook.presto.sql.tree.FrameBound.Type.UNBOUNDED_PRECEDING;
+
+public class TestSwapAdjacentWindowsByPartitionsOrder
```
please add a test when one window depends on other

maciejgrzybek replied:
https://github.com/prestodb/presto/pull/7300#discussion_r100164592

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/WindowNode.java:
```diff
@@ -90,6 +90,11 @@ public WindowNode(
         return ImmutableList.copyOf(concat(source.getOutputSymbols(), windowFunctions.keySet()));
     }
 
+    public Set<Symbol> getCreatedSymbols()
```
`getCreatedSymbols` sounds like a generic (like every `PlanNode` could have such) name while it is not a generic method as it is specified for `WindowNode`. 

I would name it as `getWindowSymbols`

maciejgrzybek replied:
Please see https://github.com/prestodb/presto/pull/7298#discussion_r100124840.
@arhimondr agrees with the proposed name.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestSwapAdjacentWindowsByPartitionsOrder.java:
```diff
@@ -104,4 +105,26 @@ public void subsetComesFirst()
     }
+
+    @Test
+    public void dependentWindowsAreNotReordered()
```
please squash this commit, and probably my previous commit (with missing test) is solved with that

maciejgrzybek replied:
It is solved. I just kept it as a fixup, to not rewrite the history while under review.

maciejgrzybek commented:
@kokosing: updated, please review changes.

maciejgrzybek commented:
Rebased on master.

---

`"reorder_windows=true"` in https://github.com/prestodb/presto/pull/7300/commits/1096572f0e89592de80bd1e9eecb4a56060d2c4c#diff-0e8c17ebc3c5d2d96616224d9062d7fbR23 is needed because of the fact we cannot use `assertUnitTest` for new rule yet because `UnaliasSymbolReferences` is not yet ported.
When it's ported, I'll change tests so we can unit-test (on DSL level) the `SwapAdjacentWindowsByPartitionsOrder` rule.
Currently we only test it using `assertPlan` tests (whole set of optimizers used).

maciejgrzybek commented:
Rebased on master.

maciejgrzybek commented:
@martint: any thoughts on this?

martint commented:
I'll take a look at this once I'm done reviewing the ORDER BY analysis PR

maciejgrzybek commented:
Rebased on master to resolve conflicts.
Removed the df87b328c9e0b73e36a5a5b19834285f83c42513 commit because @martint removed its prerequisites in 8414f0f938728d32e788b4618525945411186cbd.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:
```diff
@@ -174,7 +174,12 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
-                new ReorderWindows(), // Should be after MergeWindows to avoid unnecessary reordering of mergeable windows
+                new IterativeOptimizer(
+                        stats,
+                        ImmutableSet.of(
```
You should add ReorderWindows to the "legacy rules" argument to IterativeOptimizer. That way, it's possible to turn on the old behavior if there are any issues. Once it's been running fine for a couple of releases, we can just get rid of it.

maciejgrzybek replied:
It was there before I removed it while rebasing here: https://github.com/prestodb/presto/pull/7300#issuecomment-292247883.
The thing is `ReorderWindows` rewriter needs `PruneIdentityProjections` to work effectively but you removed that rewriter already. That's why I decided to remove it entirely because it won't be very effective anyway.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsByPartitionsOrder.java:
```diff
@@ -0,0 +1,94 @@
+
+    private static boolean dependsOn(WindowNode parent, WindowNode child)
+    {
+        return parent.getPartitionBy().stream().anyMatch(symbol -> child.getCreatedSymbols().contains(symbol))
```
> `.anyMatch(symbol -> child.getCreatedSymbols().contains(symbol))`

`.anyMatch(child.getCreatedSymbols()::contains)`

maciejgrzybek replied:
Fixed in 7982e1b728b5446376a34339de972c741d4c80a1.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsByPartitionsOrder.java:
```diff
@@ -0,0 +1,94 @@
+                || parent.getWindowFunctions().values().stream()
+                .map(WindowNode.Function::getFunctionCall)
+                .flatMap(functionCall -> functionCall.getArguments().stream())
+                .map(argument -> DependencyExtractor.extractUnique(argument))
```
DependencyExtractor::extractUnique

maciejgrzybek replied:
Fixup: 7982e1b728b5446376a34339de972c741d4c80a1.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SwapAdjacentWindowsByPartitionsOrder.java:
```diff
@@ -0,0 +1,94 @@
+                || parent.getOrderBy().stream().anyMatch(symbol -> child.getCreatedSymbols().contains(symbol))
+                || parent.getWindowFunctions().values().stream()
+                .map(WindowNode.Function::getFunctionCall)
+                .flatMap(functionCall -> functionCall.getArguments().stream())
```
Why not call DependencyExtractor.extractUnique on the FunctionCall directly (vs exploding the arguments first)?

maciejgrzybek replied:
I was not convinced how `FunctionCall` should be handled in general case in `DependencyExtractor`. But maybe that would work: c9d76c1c3bd0867fc4454345c64465aa352870e5. Then, you can apply this fixup: 1a74d1775f6da210964f34891954b359422672db.

maciejgrzybek commented:
Changed d3a0067 into f9a6ae4 per https://github.com/prestodb/presto/pull/7300#commitcomment-21685479.
Also squashed fixups mentioned before.
